features/features_test.go

```diff
@@ -378,59 +378,93 @@ func (s *featureSuite) TestAll(c *C) {
 	defer st.Unlock()
 	tr := config.NewTransaction(st)
 
+	fakeFeature := features.SnapdFeature(features.NumberOfFeatures())
+	is, why := fakeFeature.IsSupported()
+	c.Check(is, Equals, true)
+	c.Check(why, Equals, "")
+
+	restore := features.MockKnownFeaturesImpl(func() []features.SnapdFeature {
+		return []features.SnapdFeature{fakeFeature}
+	})
+	defer restore()
+
+	features.FeatureNames[fakeFeature] = "fake-feature"
+	defer func() {
+		delete(features.FeatureNames, fakeFeature)
+	}()
+
 	allFeaturesInfo := features.All(tr)
 
+	c.Assert(len(allFeaturesInfo), Equals, 1)
+
 	// Feature flags are included even if value unset
-	layoutsInfo, exists := allFeaturesInfo[features.Layouts.String()]
+	fakeFeatureInfo, exists := allFeaturesInfo[fakeFeature.String()]
 	c.Assert(exists, Equals, true)
 	// Feature flags are supported even if no callback defined.
-	c.Check(layoutsInfo.Supported, Equals, true)
+	c.Check(fakeFeatureInfo.Supported, Equals, true)
 	// Feature flags have a value even if unset.
-	c.Check(layoutsInfo.Enabled, Equals, true)
+	c.Check(fakeFeatureInfo.Enabled, Equals, false)
+
+	var fakeSupported bool
+	var fakeReason string
+	fakeCallback := func() (bool, string) {
+		return fakeSupported, fakeReason
+	}
+	features.FeaturesSupportedCallbacks[fakeFeature] = fakeCallback
+	defer func() {
+		delete(features.FeaturesSupportedCallbacks, fakeFeature)
+	}()
 
 	// Feature flags with defined supported callbacks work correctly.
 
 	// Callbacks which return false result in Supported: false
-	restore := systemd.MockSystemdVersion(229, nil)
-	defer restore()
+	fakeSupported = false
```

Comment:
I suppose that instead of having just one fake feature and manipulating the callback's return values, you could inject 2-3 features, each with different ID, each having a different name and each returning a different permutation of the values you test for, eg:

```
fakeFeature := features.SnapdFeature(features.NumberOfFeatures())
fakeFeatureUnsupported := features.SnapdFeature(features.NumberOfFeatures()+1)
fakeFeatureDisabled := features.SnapdFeature(features.NumberOfFeatures()+2)

defer features.MockFeatureNames(map[int]string{
    fakeFeature: "supported",
    fakeFeatureUnsupported: "unsupported",
    fakeFeatureDisabled: "disabled",
})()
defer features.MockFeatureSupportedCallbacks(map[string]features.SupportedFunc{
    fakeFeatureUnsupported := func() (bool, string) { return false, "reasons" },
    fakeFeatureSupportedWithComment := func() (bool, string) { return true, "reasons" },
    fakeFeatureDisabled := func() (bool, string) { 
        // supported but will be disabled in transaction
        return true, ""
    },
})()

// are the snap/conf unique for the features?

c.Assert(tr.Set("core", "experimental.unsupported", "true"), IsNil)
c.Assert(tr.Set("core", "experimental.disabled", "false"), IsNil)
...

allFeaturesInfo = features.All(tr)
c.Assert(len(allFeaturesInfo), Equals, 3)
...

unsupported, exists := allFeaturesInfo[fakeFeatureUnsupported.String()]
c.Check(unsupported.Supported, Equals, false)
c.Check(unsupported.UnsupportedReason, Matches, fakeReason)
c.Check(unsupported.Enabled, Equals, false)

disabled, exists := allFeaturesInfo[fakeFeatureUnsupported.String()]
..
```